docs: intro README (#19)
open-sir/open-sir: README.md

The current stage of the software is *Alpha*.

## Features
- Model the dynamics of infectious diseases
- Parameter fitting
- Calculation of confidence intervals
- CLI for interfacing with non Python environments (Bash, Node.JS, Matlab, etc).

So far, Open-SIR provides an implementation of the SIR model and the recently new SIR based model SIR-X, developed by the [Robert Koch Institut](http://rocs.hu-berlin.de/corona/docs/forecast/model/#sir-x-dynamics-outbreaks-with-temporally-increasing-interventions).

## Development Setup

Open-SIR uses [Pipenv](https://pipenv.pypa.io/en/latest/) to automatically create a virtual environment and manage python packages. The python packages required by Open-SIR are listed in the [Pipfile](Pipfile).

### Dependencies
* Python 3.7
* Pipenv

### Installation

After cloning the repository, change the current directory to the repository via `cd open-sir` and automatically install the environment from the Pipfile using Pipenv:

```
pipenv install
```

Next, activate the Pipenv shell:
```
pipenv shell
```

You can run the following command to check that the installation succeeded.

```

> pipenv run python open-sir.py -p '[0.95,0.38]' -i '[341555,445,0]' -t 6 > data.csv

Reviewer: #21 was merged. This can be adapted.

Author: Updated in 94e1019

case that no lockdown would be taking place.

### Command line interface
In the Pipenv shell, check that the installation was successful calling the CLI open-sir.py

```
python open-sir.py -p '[0.95,0.38]' -i '[341555,445,0]' -t 6 > data.csv
```

The output of open-sir.py is a .csv file with the predictions.

### API

You can replicate the predictions of the CLI with the following python script:

```
from models import SIR
my_SIR = SIR() # Initialize an empty SIR model
params = [0.95,0.38] # Define model parameters
w0 = [341555,445,0] # Define initial conditions
my_SIR.set_params(p=params,initial_conds=w0) # Set model parameters
n_days = 6 # Define the amount of days to predict
my_SIR.solve(n_days, n_days+1) # Call model.solve functions
sol = my_SIR.fetch() # Fetch model solution
```

### Try the Jupyter Notebook

Open and run the Jupyter Notebook [SIR-X.ipynb](SIR-X.ipynb) to:
* Get an overview of the SIR model
* Explore case studies

And learn how the API can be used to:

* Build compartmental models
* Fit parameters to existing data 
* Predict susceptible, infected and removed population
* Calculate confidence intervals of the predictions

> ### Coding style

Reviewer 2: Coding Style shouldn't be at the same level of "Usage"

Author: Yes, and also it was floating in a wrong subsection. I moved it following [PurpleBooth Template](https://gist.github.com/PurpleBooth/109311bb0361f32d87a2) below the general tests on the dev setup. Whenever we have the package, we can re-sort the README but that should be part of other PR

See f7dcb95

# open-sir

Open-SIR is an Open Source Python project for modelling pandemics and infectious diseases using Compartmental Models, such as the widely used [Susceptible-Infected-Removed (SIR) model](http://rocs.hu-berlin.de/corona/docs/forecast/model/#classic-sir-dynamics). 

The current stage of the software is *Alpha*.

## Features
- Model the dynamics of infectious diseases
- Parameter fitting
- Calculation of confidence intervals
- CLI for interfacing with non Python environments (Bash, Node.JS, Matlab, etc).

So far, Open-SIR provides an implementation of the SIR model and the recently new SIR based model SIR-X, developed by the [Robert Koch Institut](http://rocs.hu-berlin.de/corona/docs/forecast/model/#sir-x-dynamics-outbreaks-with-temporally-increasing-interventions).

## Development Setup

Open-SIR uses [Pipenv](https://pipenv.pypa.io/en/latest/) to automatically create a virtual environment and manage python packages. The python packages required by Open-SIR are listed in the [Pipfile](Pipfile).

> 

Reviewer 3: ```suggestion
```

(Extra line skip)

### Dependencies
* Python 3.7
* Pipenv

### Installation

After cloning the repository, change the current directory to the repository via `cd open-sir` and automatically install the environment from the Pipfile using Pipenv:

```
pipenv install
```

Next, activate the Pipenv shell:
```
pipenv shell
```

You can run the following command to check that the installation succeeded.

```
pipenv run start -p '[0.95,0.38]' -i '[341555,445,0]' -t 6
```

## Running the tests

Test the package running the tests out of the open-sir Pipenv environment.
```
pipenv run test
```

### Coding style tests

This project uses [Pylint](https://www.pylint.org/) and [Black 19.10b0](https://black.readthedocs.io/en/stable/) to ensure consistent coding practices, and enforced by CircleCI. Non-default parameters are available in [.pylintrc](.pylintrc).

> 

Reviewer 3: ```suggestion
```

## Usage example

You can use Open-SIR to create a 6 days prediction of the number of susceptible (S), infected (I) and removed (R) population. 
The initial conditions '-i' represent Ealing data as of 04/04/2020. The parameters '-p' provide a prediction in the hypothetical 
case that no lockdown would be taking place.

### Command line interface

In the Pipenv shell, check that the installation was successful by calling the CLI:

```
python open-sir.py -p '[0.95,0.38]' -i '[341555,445,0]' -t 6 > data.csv
```

The output of open-sir.py is a .csv file with the predictions.

*Note: On Windows, the CLI must be run from Powershell or any bash shell such as [Git BASH](https://gitforwindows.org/)*

### API

You can replicate the predictions of the CLI with the following python script:

> ```

Reviewer 3: ```suggestion
```python
```

We can get syntax highlighting that way

# open-sir

> Open source pandemics modelling API in python. The models implemented are in the scope of compartmental models of disease transmission, such as the widely used [Susceptible-Infected-Removed (SIR) model](http://rocs.hu-berlin.de/corona/docs/forecast/model/#classic-sir-dynamics). Recently, a new epidemic model based in SIR, SIR-X, was developed by the [Robert Koch Institut](http://rocs.hu-berlin.de/corona/docs/forecast/model/#sir-x-dynamics-outbreaks-with-temporally-increasing-interventions) and it is also implemented in open-sir.

Reviewer 2: If we get this README.md file before the Release, we should indicate that the current state is Alpha.

> Open source pandemics modelling API in python. The models implemented are in the scope of compartmental models of disease transmission, such as the widely used [Susceptible-Infected-Removed (SIR) model](http://rocs.hu-berlin.de/corona/docs/forecast/model/#classic-sir-dynamics). Recently, a new epidemic model based in SIR, SIR-X, was developed by the [Robert Koch Institut](http://rocs.hu-berlin.de/corona/docs/forecast/model/#sir-x-dynamics-outbreaks-with-temporally-increasing-interventions) and it is also implemented in open-sir.
> 
> The API features model construction, parameter fitting and calculation of confidence intervals.

Reviewer 2: ```suggestion
Open-SIR is an Open Source Python project for modeling pandemics using Compartmental Models, such as the widely used [Susceptible-Infected-Removed (SIR) model](http://rocs.hu-berlin.de/corona/docs/forecast/model/#classic-sir-dynamics). 

## Features
- Model the behavior of infectious diseases
- Parameter fitting
- Calculation of confidence intervals
- CLI for interfacing with non Python environments (Bash, Node.JS, Matlab, etc).

So far, Open-SIR provides an implementation of the SIR model and the recently new SIR based model SIR-X, developed by the [Robert Koch Institut](http://rocs.hu-berlin.de/corona/docs/forecast/model/#sir-x-dynamics-outbreaks-with-temporally-increasing-interventions).
```

## Getting Started

> These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See deployment for notes on how to deploy the project on a live system.

Reviewer 2: I think this should be focused in the End User. IMO it's better to distinguish between Developer documentation and User documentation

### Prerequisites

What things you need to install the software and how to install them

```
Give examples
```

### Installing

A step by step series of examples that tell you how to get a development env running

Say what the step will be

```
Give the example
```

And repeat

```
until finished
```

End with an example of getting some data out of the system or using it for a little demo

## Running the tests

Explain how to run the automated tests for this system

> ### Break down into end to end tests

Reviewer 2: I think most of the testing stuff should be in the CONTRIBUTING.md files rather than the README.md

Reviewer 2: This also include the coding style

Explain what these tests test and why

```
Give an example
```

### And coding style tests

Explain what these tests test and why

```
Give an example
```

## Deployment

> Add additional notes about how to deploy this on a live system

Reviewer 2: As described above, I think the default setup should be Deployment. We should also include the step to run Open-SIR with Docker

# open-sir

Open-SIR is an Open Source Python project for modelling pandemics and infectious diseases using Compartmental Models, such as the widely used [Susceptible-Infected-Removed (SIR) model](http://rocs.hu-berlin.de/corona/docs/forecast/model/#classic-sir-dynamics). 

The current stage of the software is *Alpha*.

## Features
- Model the dynamics of infectious diseases
- Parameter fitting
- Calculation of confidence intervals

> - CLI for interfacing with non Python environments (Bash, Node.JS, Matlab, etc).

Reviewer 3: @jia200x how would this work? I know we have a CLI, but by this statement you mean something by the lines of:

```js
exec('opensir --help')
```

?

Reviewer 2: Yes, exactly. Or someone could just generate CSV file to load them from Matlab

> So far, Open-SIR provides an implementation of the SIR model and the recently new SIR based model SIR-X, developed by the [Robert Koch Institut](http://rocs.hu-berlin.de/corona/docs/forecast/model/#sir-x-dynamics-outbreaks-with-temporally-increasing-interventions).

Reviewer 3: @jia200x @felipehuerta17 what do you think about eventually adding a new section here that lists the "upcoming" features?

I guess it is very soon to add it now, but maybe next week once we have those more clear?

Reviewer 2: +1!

Author: Done. Let's include it now as it can attract interest!
We can "commit" to at least two approaches in the future

* Incorporate SEIR model
* Advanced methods to estimate reproduction number

> ## Getting Started

Reviewer 3: ```suggestion
## Development Setup
```

I think that title is a better fit for this section, because it is intended for devs.

Once we have a CLI, I would add a `Quickstart` section before this one.

What do you all think?

Reviewer 2: I agree. Everything that is `pipenv` related should be reserved for developers.
We will add a proper quickstart as soon as we create the package, so we could got with this and update later as needed

Open-SIR uses [Pipenv](https://pipenv.pypa.io/en/latest/) to automatically create a virtual environment and manage python packages. The python packages required by Open-SIR are listed in the [Pipfile](Pipfile).

> ### Dependencies

Reviewer 3: I am not sure about this section. It is a list which could easily get out of date every time someone adds a new dependency to the `Pipfile` and forgets to add it here.

I think it is important to say that you need `Python 3.7` and `Pipenv`, but about the rest of the libraries I am not so sure.

Reviewer 2: I agree with this. Considering it's just a developer environment, both dependencies should suffice.

Author: But what if somebody wants to use open-sir before we make it a package?

Reviewer 2: But it can still use it with "pipenv"

Reviewer 3: Yes. Knowing that it uses `Pipenv` for dependency management should be enough. At no moment should that dev manually add the dependencies. All is automatically sorted with `pipenv install`.

Author: OK. Implemented

* Python 3.7
* Pipenv
* NumPy
* SciPy
* Jupyter
* Matplotlib

> ### Installation

Reviewer 3: Should we add a "docs generation" subsection to this section?

Reviewer 2: Yes! Specially since we don't have a "readthedocs" entry for now :)

> After cloning the repository, change the current directory to the repository by `cd open-sir` and automatically install the environment from the Pipfile using Pipenv:

Reviewer 3: ```suggestion
After cloning the repository, change the current directory to the repository via `cd open-sir` and automatically install the environment from the Pipfile using Pipenv:
```

```
pipenv install
```

> Next, activate the Pipenv shell:

Reviewer 3: Strictly speaking, activating this shell is not part of the installation. I would add this part to the next section, where the CLI is used inside the shell.

```
pipenv shell
```

> ## Usage example

Author: I think I understand what you get in mind. I changed the layout of the document.

Examples
-- CLI
-- API
-- Jupyter Notebook

I commited a new version with everything except the contributors

> ### Command line interface

Reviewer 3: ```suggestion
### Command line interface

```

> In the Pipenv shell, check that the installation was successful calling the CLI open-sir.py

Reviewer 3: ```suggestion
In the Pipenv shell, check that the installation was successful by calling the CLI:
```

python open-sir.py -p '[0.95,0.38]' -i '[341555,445,0]' -t 6 
```

> Alternatively, run the same command through Pipenv outside the open-sir Pipenv environment:

Reviewer: ```suggestion
Alternatively, instead of activating the shell, you can run the same command through Pipenv outside the open-sir Pipenv environment:
```

```

> pipenv python open-sir.py -p '[0.95,0.38]' -i '[341555,445,0]' -t 6 

Reviewer: Just a heads-up, this can change to `pipenv run start -p '[0.95,0.38]' -i '[341555,445,0]' -t 6` with #21.

Author: Thanks @leandrolanzieri , should I provide both options now? I will keep an eye on #21. Many thanks for the contribution!

I also realized that I missed the run before python. It is updated in a commit now.

> ```

Author: Although in the Jupyter Notebook the API is used extensively, I understand that we should add some example snippets! that would be much better.

With respect to your second command, I understand but it would be weird if it appears before the usage. Should we remove it completely?

> ```

Author: @jia200x @sasalatart these changes were implemented in the latest version on the readme

> ```

Author: 9385dba

pipenv run python open-sir.py -p '[0.95,0.38]' -i '[341555,445,0]' -t 6 
```

The output of open-sir.py is a table with a 6 day prediction of the number of susceptible (S), infected (I) and removed (R) population. The initial conditions -i represent Ealing data as of 04/04/2020. The parameters provide a prediction in the hypothetical case that no lockdown is taking place.

> The last line of the output should be

Author: Unfortunately CLI is not working on Windows CMD, so depending on our timelines for the release we may have to specify that CLI is only working in sh and bash.  See #30

> The last line of the output should be

Author: As this can run in Windows Powershell, I would carry on with your suggestion

> The last line of the output should be

Author: All implemented in 7da9602

> ```

Reviewer: This could be moved inside `Usage example`

Author: Sorted

6.000000000000000000e+00,3.213056091837603017e+05,1.233370563811089232e+04,8.360685178128811458e+03

> ```
> ### Python API usage

Reviewer 3: ```suggestion
```

### Python API usage
```

Reviewer 3: Also, should this say `Jupyter Notebook` instead? I think it is more precise

Open and run the Jupyter Notebook [SIR-X.ipynb](SIR-X.ipynb) to:
* Get an overview of the SIR model
* Explore case studies

And learn how the API can be used to:

* Build compartmental models
* Fit parameters to existing data 
* Predict susceptible, infected and removed population
* Calculate confidence intervals of the predictions

> 

Reviewer 3: ```suggestion
```

## Running the tests

> Test the package running the tests out of the open-sir Pipenv environment.

Reviewer 3: ```suggestion
You may run the tests by executing the following command:
```

Author: @leandrolanzieri initially changed the first wording to this second. Do you agree with the change proposed by @sasalatart ?

Reviewer: The latest version looks fine!

Author: Updated!

```
pipenv run test
```

### Coding style

> Coding analysis is automatically reviewed using [Pylint](https://www.pylint.org/) and [Black 19.10b0](https://black.readthedocs.io/en/stable/) with the non-default parameters available in [.pylintrc](pylintrc).

Reviewer 3: ```suggestion
This project uses [Pylint](https://www.pylint.org/) and [Black 19.10b0](https://black.readthedocs.io/en/stable/) to ensure consistent coding practices, and enforced by CircleCI. Non-default parameters are available in [.pylintrc](pylintrc).
```

Author: Great, thanks!

> ## Authors

Reviewer 3: @felipehuerta17 @jia200x take a look at this repo:

https://github.com/all-contributors/all-contributors/blob/master/README.md

I think it is a very nice way of specifying contributors

Author: Amazing! I will setup this today

Reviewer 2: let's use that one! It's really nice

Author: Alright, this will be setup after this PR as you can see in #29, this is performed automatically using the bot and rewrites an existing readme.

> * **José Álamos** - [RIOT](https://github.com/RIOT-OS)

Reviewer 3: Maybe we can specify GitHub usernames here instead. What do you think?

Reviewer 2: +1

Author: @sasalatart  Do you mean after the name of the authors?

Reviewer 3: Hmmm or maybe instead of the names. Not sure though

Author: Let's include the link of the github on our names.

* **Felipe Huerta** - [PhD Student](https://www.imperial.ac.uk/people/f.huerta-perez17) at [Imperial College London](https://github.com/ImperialCollegeLondon)
* **Sebastián Salata** - Software Engineer - Full Stack

See also the list of [contributors](https://github.com/open-sir/open-sir/contributors) who participated in this project.

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details

> ## Acknowledgements

Reviewer 3: 👍 I love this ❤️
